File: app/src/main/java/com/sdin/tourstamprally/ui/fragment/DirectionGuidFragment.java

package com.sdin.tourstamprally.ui.fragment;

import android.annotation.SuppressLint;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.navigation.Navigation;

import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;

import com.sdin.tourstamprally.R;
import com.sdin.tourstamprally.adapter.DirectionGuid_Adapter;
import com.sdin.tourstamprally.adapter.DirectionGuid_Tag_Adapter;
import com.sdin.tourstamprally.adapter.SelectLocationAdapter;
import com.sdin.tourstamprally.databinding.FragmentDirectionGuidBinding;
import com.sdin.tourstamprally.model.Location_four;
import com.sdin.tourstamprally.model.TourTagModel;
import com.sdin.tourstamprally.v2.RecyclerViewListener;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class DirectionGuidFragment extends BaseFragment implements RecyclerViewListener {
    //길안내 관광지

    private static final String ARG_PARAM = "model";
    private FragmentDirectionGuidBinding binding;
    //private List<Tour_Spot> tourList;

    private DirectionGuid_Adapter adapter;
    private DirectionGuid_Tag_Adapter tagAdapter;
    // 총 참여자

    private ArrayList<Location_four> paramArrayList;
    private final ArrayList<TourTagModel> hashTagList = new ArrayList<>();

    public DirectionGuidFragment() {

    }

    public static DirectionGuidFragment newInstance(ArrayList<Location_four> param1) {
        DirectionGuidFragment fragment = new DirectionGuidFragment();
        Bundle args = new Bundle();
        args.putParcelableArrayList(ARG_PARAM, param1);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            paramArrayList = getArguments().getParcelableArrayList(ARG_PARAM);
            Log.wtf("넘어옴?", String.valueOf(paramArrayList.size()));
            /*Log.wtf("paramArrayList", paramArrayList.toString());
            Log.wtf("paramArrayListsize", String.valueOf(paramArrayList.size()));*/
        } else {
            Log.wtf("넘어옴 ㄴㄴ?", "0");
        }
    }


    private void getData() {
        binding.directionGuidPgb.setVisibility(View.VISIBLE);

        apiService.getHashTag().enqueue(new Callback<List<TourTagModel>>() {
            @Override
            public void onResponse(@NotNull Call<List<TourTagModel>> call, @NotNull Response<List<TourTagModel>> response) {
                try {
                    if (response.isSuccessful()) {
                        List<TourTagModel> hashTagModels = response.body();

                        if (hashTagModels != null) {

                            for (TourTagModel model : hashTagModels) {
                                if (model == null) {
                                    Log.wtf("TourTagModel", "NULL!!");
                                }
                                if (model != null && model.getHashTag() != null) {
                                    String[] array = Arrays.stream(model.getHashTag().split("#")).map(String::trim).toArray(String[]::new);
                                    Arrays.stream(array)
                                            .filter(s -> (s != null && s.length() > 0))
                                            .distinct()
                                            .forEach(data -> hashTagList.add(new TourTagModel(data, model.getLocation_idx())));
                                }


                            }
                        }

                        setHashTag();
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }

            }

            @Override
            public void onFailure(@NotNull Call<List<TourTagModel>> call, @NotNull Throwable t) {
                t.printStackTrace();
            }
        });


    }

    private void setTourSpotList() {

        adapter = new DirectionGuid_Adapter(
                paramArrayList,
                this,
                requireContext()
                /*requireActivity()*//*,
                location_Progress_Map,
                location_history_Map,
                particiMap*/);

        binding.locationRe.setAdapter(adapter);
        binding.locationRe.setHasFixedSize(true);
    }


    private void setHashTag() {

        tagAdapter = new DirectionGuid_Tag_Adapter(hashTagList);
        tagAdapter.setOnItemClickListener(param -> {
            //Log.wtf("setOnItemClickListener", "param = " + param +"1");
            search(param.getHashTag());
        });
        binding.tagRe.setAdapter(tagAdapter);
        binding.tagRe.setHasFixedSize(true);
        binding.directionGuidPgb.setVisibility(View.GONE);
    }


    @Override
    public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_direction_guid, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstancdState) {
        SelectLocationAdapter spinnerAdapter = new SelectLocationAdapter(requireContext(), Arrays.asList(requireContext().getResources().getStringArray(R.array.area_direction)));
        binding.spinnerTourRecord.setAdapter(spinnerAdapter);
        binding.spinnerTourRecord.setOnItemSelectedListener(selectedListener);
        setTourSpotList();
        getData();
        binding.searchEdt.addTextChangedListener(new TextWatcher() {

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                search(binding.searchEdt.getText().toString());
            }

        });

        binding.searchBtn.setOnClickListener(v -> search(binding.searchEdt.getText().toString()));
    }

    @SuppressLint("NotifyDataSetChanged")
    private void search(String searchData) {

        ArrayList<Location_four> arrayList = new ArrayList<>();

        Set<Integer> locationIdxes = new HashSet<>();

        if (searchData.length() == 0) {


            arrayList.addAll(paramArrayList);
            adapter.setList(arrayList);

        } else {

            for (TourTagModel tagModel : hashTagList) {
                if (tagModel.getHashTag().toLowerCase().contains(searchData)) {
                    locationIdxes.add(tagModel.getLocation_idx());
                }
            }

            for (Location_four location_model : paramArrayList) {
                if (location_model.getLocation_name().toLowerCase().equals(searchData)) {
                    locationIdxes.add(location_model.getLocation_idx());
                }
            }


            for (Integer idx : locationIdxes) {
                for (Location_four location_model : paramArrayList) {
                    if (idx == location_model.getLocation_idx()) {
                        arrayList.add(location_model);
                    }
                }
            }

        }


        adapter.setList(arrayList);
        adapter.notifyDataSetChanged();
    }


    private final AdapterView.OnItemSelectedListener selectedListener = new AdapterView.OnItemSelectedListener() {
        @Override
        public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
            //Log.d("getSelectedItem", parent.getSelectedItem().toString());

            if (paramArrayList != null) {
                String searchData = parent.getSelectedItem().toString();
                if (searchData.equals("전체"))
                    searchData = "";
                search(searchData);
            }

        }

        @Override
        public void onNothingSelected(AdapterView<?> parent) {

        }
    };


    @Override
    public void onLocationItemClick(@NonNull Location_four location_four) {
        Bundle bundle = new Bundle();
        bundle.putParcelable("model", location_four);
        bundle.putString("title", location_four.getLocation_name() + " 랠리 맵");
        Navigation.findNavController(requireActivity(), R.id.nav_host)
                .navigate(R.id.action_fragment_direction_guid_to_fragment_location, bundle);



    }
}

File: app/src/main/java/com/sdin/tourstamprally/ui/fragment/tour/DirectionGuidFragment.kt
package com.sdin.tourstamprally.ui.fragment.tour

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.findNavController
import androidx.navigation.fragment.navArgs
import androidx.viewpager2.adapter.FragmentStateAdapter
import com.google.android.material.tabs.TabLayoutMediator
import com.sdin.tourstamprally.R
import com.sdin.tourstamprally.databinding.FragmentDirectionGuidBinding
import com.sdin.tourstamprally.model.TopFourLocationModel
import com.sdin.tourstamprally.model.TourTagModelDC
import com.sdin.tourstamprally.ui.fragment.BaseFragment
import com.sdin.tourstamprally.ui.fragment.report.review.MoreReviewFragment
import com.sdin.tourstamprally.utill.listener.Observer
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class DirectionGuidFragment : BaseFragment(), com.sdin.tourstamprally.utill.listener.Observable {

    private var binding: FragmentDirectionGuidBinding? = null
    private val args: DirectionGuidFragmentArgs by navArgs()
    private lateinit var paramList: MutableList<TopFourLocationModel>
    private val hashTagList = mutableListOf<TourTagModelDC>()
    //private lateinit var directionGuidAdapter: DirectionGuidAdapter

    private val observerList = arrayListOf<Observer>()
    private val guidFragment by lazy {
        DirectionGuid2Fragment().newInstance(paramList)
        //DirectionGuid2Fragment.newInstance(paramList)
    }

    private val moreReviewFrag by lazy {
        MoreReviewFragment()
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding =
            DataBindingUtil.inflate(inflater, R.layout.fragment_direction_guid, container, false)

        binding?.fragment = this

        binding?.btnGetstamp?.setOnClickListener {
            val action = DirectionGuidFragmentDirections.actionFragmentDirectionGuidToPageStamp()
            findNavController().navigate(action)
        }




        return binding?.root
    }

    override fun onViewCreated(view: View, savedInstancdState: Bundle?) {
        super.onViewCreated(view, savedInstancdState)
        paramList = args.locationModelList.toMutableList()

        initView()
        getData()


    }
    inner class ViewPagerAdapter(private val list : MutableList<Fragment>)
        : FragmentStateAdapter(requireActivity()) {

        override fun getItemCount(): Int {
            return list.count()
        }

        override fun createFragment(position: Int): Fragment {
            return list[position]
        }

    }

    private fun initView() {


        binding?.apply {

            //viewpager
            val fragmentList = mutableListOf<Fragment>()
            //val guidFragment = DirectionGuid2Fragment.newInstance(paramList)
            fragmentList.add(guidFragment)
            fragmentList.add(moreReviewFrag)
            registerObserver(guidFragment)
            registerObserver(moreReviewFrag)


            viewpagerguide.adapter = ViewPagerAdapter(fragmentList)
            TabLayoutMediator(binding?.tabLayout!!, binding?.viewpagerguide!!){ tab, position ->
                when(position){
                    0 -> tab.text = "투어"
                    1 -> tab.text = "리뷰"
                }

            }.attach()

            searchEdt.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(
                    s: CharSequence?,
                    start: Int,
                    count: Int,
                    after: Int
                ) = Unit

                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) =
                    Unit

                override fun afterTextChanged(s: Editable?) {
                    search(searchData = searchEdt.text.toString())
                }

            })



            searchBtn.setOnClickListener {
                search(searchData = searchEdt.text.toString())
            }

            /*locationRe.apply {

                setHasFixedSize(true)
                layoutManager = LinearLayoutManager(requireContext())

                directionGuidAdapter = DirectionGuidAdapter(requireContext()) { locationModel ->
                    val action =
                        DirectionGuidFragmentDirections.actionFragmentDirectionGuidToFragmentLocation(
                            locationModel.location_name + " 랠리 맵",
                            locationModel
                        )


                    if (locationModel.touristspotpoint_createtime == null) {
                        Toast.makeText(requireContext(), "등록된 관광지가 없습니다.", Toast.LENGTH_SHORT)
                            .show()
                        return@DirectionGuidAdapter
                    }

                    findNavController().navigate(action)


                }.apply {
                    submitList(paramList)
                }

                adapter = directionGuidAdapter
            }*/
        }
    }

    private fun getData() {
//        binding?.directionGuidPgb?.visibility = View.VISIBLE

        apiService.hashTag.enqueue(object : Callback<List<TourTagModelDC>> {
            override fun onResponse(
                call: Call<List<TourTagModelDC>>,
                response: Response<List<TourTagModelDC>>
            ) {
                if (response.isSuccessful) {
                    val result = response.body()

                    if (!result.isNullOrEmpty()) {
                        for (model in result) {
                            if (model != null) {
                                model.hashTag?.let {
                                    val arr: List<String> = it.split("#")
                                    val set = arr.toSet()


                                    for ((index, str) in set.withIndex()) {
                                        //Log.wtf("str $index", str)
                                        if (str.trim() != "") {
                                            hashTagList.add(TourTagModelDC(str, model.location_idx))
                                        }

                                    }

                                    /*val strArr: Array<List<String>> = arrayOf(model.hashTag.split("#"))
                                    if (strArr != null) {

                                        for ((index, str) in strArr.withIndex()) {
                                            Log.wtf("srt $index", str[index])
                                            hashTagList.add(
                                                TourTagModelDC(
                                                    str[index],
                                                    model.location_idx
                                                )
                                            )
                                        }
                                    }*/
                                }
                            }
                        }

                        //setHashTag()
                    }
                }
            }

            override fun onFailure(call: Call<List<TourTagModelDC>>, t: Throwable) {
                t.printStackTrace()
            }

        })
    }

    private fun setHashTag() {
        if (context == null) {
            return
        }

        /*binding?.tagRe?.apply {
            adapter = DirectionGuidTagAdapter() { model ->
                model.hashTag?.let { search(it) }
            }.apply {
                submitList(hashTagList)
            }

            setHasFixedSize(true)
            layoutManager =
                LinearLayoutManager(requireContext(), LinearLayoutManager.HORIZONTAL, true)
        }

        binding?.directionGuidPgb?.visibility = View.GONE*/
    }

    private val selectedListener = object : AdapterView.OnItemSelectedListener {
        override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
            if (paramList != null) {
                var searchData = parent?.selectedItem.toString()
                if (searchData == "전체") {
                    searchData = ""
                }

                search(searchData)
            }
        }

        override fun onNothingSelected(parent: AdapterView<*>?) = Unit

    }

    /*class Observerparent{
        val parentObserver = mutableListOf<(String)->Unit>()
        var observerdata : String by Delegates.observable(""){property, oldValue, newValue ->
            parentObserver.forEach { it(newValue) }

        }
    }*/

    private fun search(searchData: String) {
        val searchList = mutableListOf<TopFourLocationModel>()

        if (searchData.isEmpty()) {
            searchList.addAll(paramList)

        } else {

            notifyObservers(searchData)
        }

    }


    companion object {

        @JvmStatic
        fun newInstance() =
            MoreReviewFragment()
    }

    override fun registerObserver(observer: Observer) {
        observerList.add(observer)
    }

    override fun unregisterObserver(observer: Observer) {
        observerList.remove(observer)
    }

    override fun notifyObservers(search: String) {
        observerList.forEach {
            it.update(search)

        }
    }


}